### PR TITLE
Presentation: Unique values in instance filter should only be retrieved from instances that are child nodes for filtered parent node

### PR DESCRIPTION
**Pull request by medikornov**
Closes: https://github.com/iTwin/itwinjs-core/issues/5725
itwinjs-core: https://github.com/iTwin/itwinjs-core/tree/presentation/unique-values-based-on-parent-node
itwinjs-core PR: https://github.com/iTwin/itwinjs-core/pull/5763

### CODE REVIEW

File: iModelCore/ECPresentation/PublicAPI/ECPresentation/Content.h [hunk 1]
```diff
@@ -1113,6 +1113,7 @@ struct EXPORT_VTABLE_ATTRIBUTE ContentDescriptor : RefCountedBase
 
 private:
     PresentationRuleSetCPtr m_ruleset;
+    PresentationRuleSetPtr m_hierarchyLevelRuleset;
```
Comment (grigasp) [inline iModelCore/ECPresentation/PublicAPI/ECPresentation/Content.h:1116]: is this ruleset different from `m_ruleset`?

File: iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp [hunk 2]
```diff
@@ -1198,7 +1198,9 @@ ContentDescriptorCPtr RulesDrivenECPresentationManagerImpl::_GetNodesDescriptor(
             ),
         *KeySet::Create(params.GetParentNodeKey() ? NavNodeKeyList{ params.GetParentNodeKey() } : NavNodeKeyList{})
         ), params);
-    return GetContentDescriptor(descriptorParams);
+    auto descriptor = GetContentDescriptor(descriptorParams);
+    (const_cast<ContentDescriptor*>(descriptor.get()))->SetHierarchyLevelRuleset(*ruleset);
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp:1202]: This is not good. The descriptor we get from `GetContentDescriptor` possibly comes from cache, so we definitely don't want to mutate it. Instead, we could clone it using `ContentDescriptor::Create(*sourceDescriptor)` and then modify.

**Review by grigasp — CHANGES_REQUESTED** [review]

File: iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp (hunk 2 above)
Comment (medikornov) [reply iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp]: fixed

File: iModelCore/ECPresentation/PublicAPI/ECPresentation/Content.h [hunk 3]
```diff
@@ -1285,6 +1288,11 @@ struct EXPORT_VTABLE_ATTRIBUTE ContentDescriptor : RefCountedBase
     //! Should only distinct values be returned
     bool OnlyDistinctValues() const {return HasContentFlag(ContentFlags::DistinctValues);}
 #endif
+
+    //! Set if this descriptor is created using a different from the input ruleset
+    void SetDifferentFromInputRuleset(bool value) {m_differentFromInputRuleset = value;}
+    //! Is the descriptor created using a different from the input ruleset
+    bool IsDifferentFromInputRuleset() const {return m_differentFromInputRuleset;}
```
Comment (grigasp) [inline iModelCore/ECPresentation/PublicAPI/ECPresentation/Content.h:1295]: maybe just set this to `true` in `SetRuleset`?

File: iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp [hunk 4]
```diff
@@ -1198,7 +1198,13 @@ ContentDescriptorCPtr RulesDrivenECPresentationManagerImpl::_GetNodesDescriptor(
             ),
         *KeySet::Create(params.GetParentNodeKey() ? NavNodeKeyList{ params.GetParentNodeKey() } : NavNodeKeyList{})
         ), params);
-    return GetContentDescriptor(descriptorParams);
+    auto descriptor = ContentDescriptor::Create(*GetContentDescriptor(descriptorParams));
+    if (descriptor.IsValid())
```
Comment (grigasp) [inline iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp:1202]: `ContentDescriptor::Create` will always return a descriptor, but `GetContentDescriptor(descriptorParams)` is not guaranteed to. This should be changed to:
```cpp
auto descriptor = GetContentDescriptor(descriptorParams);
if (descriptor.IsNull())
    return nullptr;

auto result = ContentDescriptor::Create(*descriptor);
...
```

**Review by grigasp — CHANGES_REQUESTED** [review]

File: iModelCore/ECPresentation/PublicAPI/ECPresentation/Content.h (hunk 3 above)
Comment (grigasp) [reply iModelCore/ECPresentation/PublicAPI/ECPresentation/Content.h:1295]: Also, `IsDifferentFromInputRuleset` sounds like we're comparing descriptor with a ruleset. Maybe rename to `UsesModifiedRuleset`?

Comment (medikornov) [reply iModelCore/ECPresentation/PublicAPI/ECPresentation/Content.h]: fixed

File: iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp (hunk 4 above)
Comment (medikornov) [reply iModelCore/ECPresentation/Source/PresentationManagerImpl.cpp]: fixed

**Review by grigasp — APPROVED** [review]

**Pull request merged**